### PR TITLE
Implement proper security client http codes

### PR DESCRIPTION
**Pull request by MatthijsKrempel**
PR for https://github.com/OrchardCMS/OrchardCore/issues/3209

Replaced all UnauthorizedResult with ForbiddenResult on controllers that have the AdminAttribute or the AuthorizeAttribute.
Created extensionmethod that returns ChallengeResult or ForbiddenResult depending on wether the user is authenticated.
Replaced all UnauthorizeResult on controllers that support anonymous to this extension method.

### CODE REVIEW

**Comment by jtkech** [comment]
> Replaced all UnauthorizedResult with ForbiddenResult on controllers that have the AdminAttribute  or the AuthorizeAttribute

OR maybe whose controller name (without the `Controller` suffix) is equal to `Admin` without having to add the `[Admin]` attribute. This because our admin filters do the same for controllers that have the `[Admin]` attribute or whose controller name is equal to `Admin`

File: src/OrchardCore.Modules/OrchardCore.Features/Controllers/AdminController.cs [hunk 1]
```diff
@@ -19,6 +19,7 @@
 
 namespace OrchardCore.Features.Controllers
 {
+    [Admin]
     public class AdminController : Controller
```
Comment (deanmarcussen) [inline src/OrchardCore.Modules/OrchardCore.Features/Controllers/AdminController.cs]: Looks good @MatthijsKrempel but for the extra `Admin` attributes.

We have a well defined (if misunderstood) convention that it's an Admin controller if called `AdminController` or if has `[Admin]`.

I don't think we need to muddy the waters anymore by then putting the attribute on controllers which conform to the convention.

Also you only added it to some admin controllers, so more confusing ;)

Maybe I try and write some docs for it

Comment (MatthijsKrempel) [reply src/OrchardCore.Modules/OrchardCore.Features/Controllers/AdminController.cs]: I can remove those if you want, I think it is more explicit to use the attribute rather then rely on a convention. But up to you guys.

**Comment by MatthijsKrempel** [comment]
@jtkech @deanmarcussen removed all [Admin] attributes from controllers named AdminController.

**Comment by sebastienros** [comment]
@kevinchalet @TFleury would you mind confirming this is the correct fix?

File: src/OrchardCore.Modules/OrchardCore.Users/Controllers/TestController.cs [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+using OrchardCore.Users.Workflows.Activities;
+using OrchardCore.Workflows.Services;
+
+namespace OrchardCore.Users.Controllers
+{
+    [ApiController]
+    [IgnoreAntiforgeryToken]
+    [AllowAnonymous]
+    [Route("Test")]
+    public class TestController : ControllerBase
```
Comment (TFleury) [inline src/OrchardCore.Modules/OrchardCore.Users/Controllers/TestController.cs]: Please remove this file

File: src/OrchardCore.Modules/OrchardCore.ContentLocalization/Controllers/AdminController.cs [hunk 3]
```diff
@@ -48,7 +49,7 @@ public async Task<IActionResult> Localize(string contentItemId, string targetCul
 
             if (!await _authorizationService.AuthorizeAsync(User, Permissions.LocalizeContent, contentItem))
             {
-                return Unauthorized();
+                return this.Forbid();
```
Comment (TFleury) [inline src/OrchardCore.Modules/OrchardCore.ContentLocalization/Controllers/AdminController.cs]: Remove `this.`

File: src/OrchardCore.Modules/OrchardCore.ContentLocalization/Controllers/AdminController.cs [hunk 4]
```diff
@@ -58,7 +59,7 @@ public async Task<IActionResult> Localize(string contentItemId, string targetCul
 
             if (!await _authorizationService.AuthorizeAsync(User, CommonPermissions.EditContent, checkContentItem))
             {
-                return Unauthorized();
+                return this.ChallengeOrForbid();
```
Comment (TFleury) [inline src/OrchardCore.Modules/OrchardCore.ContentLocalization/Controllers/AdminController.cs]: Admin controller : `return Forbid();`

File: src/OrchardCore.Modules/OrchardCore.ContentLocalization/Controllers/AdminController.cs [hunk 5]
```diff
@@ -8,6 +8,7 @@
 using OrchardCore.Contents;
 using OrchardCore.DisplayManagement.ModelBinding;
 using OrchardCore.DisplayManagement.Notify;
+using OrchardCore.Mvc.Utilities;
```
Comment (TFleury) [inline src/OrchardCore.Modules/OrchardCore.ContentLocalization/Controllers/AdminController.cs]: Not required

File: src/OrchardCore.Modules/OrchardCore.Contents/Controllers/ApiController.cs [hunk 6]
```diff
@@ -70,7 +71,7 @@ public async Task<IActionResult> Post(ContentItem newContentItem, bool draft = f
             {
                 if (!await _authorizationService.AuthorizeAsync(User, Permissions.PublishContent))
                 {
-                    return Unauthorized();
+                    return Forbid();
```
Comment (TFleury) [inline src/OrchardCore.Modules/OrchardCore.Contents/Controllers/ApiController.cs]: Maybe anonymous : `return this.ChallengeOrForbid();`

File: src/OrchardCore.Modules/OrchardCore.Contents/Controllers/ApiController.cs [hunk 7]
```diff
@@ -81,7 +82,7 @@ public async Task<IActionResult> Post(ContentItem newContentItem, bool draft = f
             {
                 if (!await _authorizationService.AuthorizeAsync(User, Permissions.EditContent, contentItem))
                 {
-                    return Unauthorized();
+                    return Forbid();
```
Comment (TFleury) [inline src/OrchardCore.Modules/OrchardCore.Contents/Controllers/ApiController.cs]: Maybe anonymous : `return this.ChallengeOrForbid();`

File: src/OrchardCore.Modules/OrchardCore.Resources/package-lock.json [hunk 8]
```diff
@@ -4,6 +4,11 @@
   "lockfileVersion": 1,
   "requires": true,
   "dependencies": {
+    "@fortawesome/fontawesome-free": {
```
Comment (TFleury) [inline src/OrchardCore.Modules/OrchardCore.Resources/package-lock.json]: Changes in this file are not related to the PR.

**Comment by kevinchalet** [comment]
@sebastienros it's likely good enough in our case, but the perfect fix would involve a bit more code if we wanted to support `IPolicyEvaluator`. Here's what MVC does: https://github.com/dotnet/aspnetcore/blob/master/src/Mvc/Mvc.Core/src/Authorization/AuthorizeFilter.cs#L186-L205

**Comment by MatthijsKrempel** [comment]
So, if we could tie the permissions framework into the policy framework, we would have a really nice feature.
Will process the code reviews and have a look.

**Comment by MatthijsKrempel** [comment]
@TFleury thank you for the review!

File: src/OrchardCore.Modules/OrchardCore.Lucene/Controllers/SearchController.cs [hunk 9]
```diff
@@ -67,8 +68,14 @@ ILogger<SearchController> logger
         [HttpGet]
         public async Task<IActionResult> Search(SearchIndexViewModel viewModel, PagerSlimParameters pagerParameters)
         {
+            if (!await _authorizationService.AuthorizeAsync(User, Permissions.QueryLuceneApi))
+            {
+                return this.ChallengeOrForbid();
```
Comment (TFleury) [inline src/OrchardCore.Modules/OrchardCore.Lucene/Controllers/SearchController.cs]: Move to line 79

Comment (MatthijsKrempel) [reply src/OrchardCore.Modules/OrchardCore.Lucene/Controllers/SearchController.cs]: @TFleury not sure if it is correct, can you check?
